File: hw/pastor/u3/hw35CrudStrategy/CrudStrategyPY/controller/customer_controller.py
from model.crud_context import CrudContext
from model.customer import Customer
from controller.validation_controller import ValidationController
from utils.id_generator import IdGenerator
from typing import List

class CustomerController:
    def __init__(self, storage_type: str = "JSON"):
        self.crud_context = CrudContext(storage_type)
    
    def set_storage_type(self, storage_type: str):
        self.crud_context.set_storage_strategy(storage_type)
    
    def get_current_storage_type(self) -> str:
        return self.crud_context.get_current_storage_type()
    
    def create_customer(self, name: str, apartment_number: str, email: str, phone: str) -> Customer:
        validation_errors = ValidationController.validate_all_fields(
            name, apartment_number, email, phone)
        
        if validation_errors:
            raise ValueError(f"Validation errors:\n{validation_errors}")
        
        customer_id = IdGenerator.generate_unique_id()
        
        return Customer(customer_id, name, apartment_number, email, phone)
    
    def add_customer(self, customer: Customer) -> bool:
        return self.crud_context.add_customer(customer)
    
    def add_customer_from_fields(self, name: str, apartment_number: str, email: str, phone: str) -> bool:
        try:
            customer = self.create_customer(name, apartment_number, email, phone)
            return self.add_customer(customer)
        except ValueError as e:
            raise e
        except Exception as e:
            print(f"Error adding customer: {e}")
            return False
    
    def delete_customer(self, customer_id: int) -> bool:
        return self.crud_context.delete_customer(customer_id)
    
    def update_customer(self, customer_id: int, name: str, apartment_number: str, 
                       email: str, phone: str) -> bool:
        try:
            validation_errors = ValidationController.validate_all_fields(
                name, apartment_number, email, phone)
            
            if validation_errors:
                raise ValueError(f"Validation errors:\n{validation_errors}")
            
            updated_customer = Customer(customer_id, name, apartment_number, email, phone)
            return self.crud_context.update_customer(customer_id, updated_customer)
        except ValueError as e:
            raise e
        except Exception as e:
            print(f"Error updating customer: {e}")
            return False
    
    def get_all_customers(self) -> List[Customer]:
        return self.crud_context.get_all_customers()
    
    def get_customer_by_id(self, customer_id: int) -> Customer:
        return self.crud_context.get_customer_by_id(customer_id)
    
    def get_all_customers_formatted(self) -> str:
        customers = self.get_all_customers()
        result = f"=== CUSTOMERS ({self.get_current_storage_type()}) ===\n"
        
        for customer in customers:
            result += f"{customer}\n"
        
        result += f"Total: {len(customers)} customers\n"
        return result

File: hw/aymacana/u3/hw35CrudStrategy/CrudStrategyJS/controllers/CustomerController.js
import { CrudContext } from '../models/CrudContext.js';
import { Customer } from '../models/Customer.js';
import { ValidationController } from './ValidationController.js';
import { IdGenerator } from '../utils/IdGenerator.js';

export class CustomerController {
    constructor(storageType = "JSON") {
        this.crudContext = new CrudContext(storageType);
    }

    setStorageType(storageType) {
        this.crudContext.setStorageStrategy(storageType);
    }

    getCurrentStorageType() {
        return this.crudContext.getCurrentStorageType();
    }

    async createCustomer(name, apartmentNumber, email, phone) {
        const validationErrors = ValidationController.validateAllFields(
            name, apartmentNumber, email, phone
        );
        
        if (validationErrors) {
            throw new Error(`Validation errors:\n${validationErrors}`);
        }
        
        const customerId = await IdGenerator.generateUniqueId();
        return new Customer(customerId, name, apartmentNumber, email, phone);
    }

    async addCustomer(customer) {
        return await this.crudContext.addCustomer(customer);
    }

    async addCustomerFromFields(name, apartmentNumber, email, phone) {
        try {
            const customer = await this.createCustomer(name, apartmentNumber, email, phone);
            return await this.addCustomer(customer);
        } catch (error) {
            throw error;
        }
    }

    async deleteCustomer(customerId) {
        return await this.crudContext.deleteCustomer(customerId);
    }

    async updateCustomer(customerId, name, apartmentNumber, email, phone) {
        try {
            const validationErrors = ValidationController.validateAllFields(
                name, apartmentNumber, email, phone
            );
            
            if (validationErrors) {
                throw new Error(`Validation errors:\n${validationErrors}`);
            }
            
            const updatedCustomer = new Customer(customerId, name, apartmentNumber, email, phone);
            return await this.crudContext.updateCustomer(customerId, updatedCustomer);
        } catch (error) {
            throw error;
        }
    }

    async getAllCustomers() {
        return await this.crudContext.getAllCustomers();
    }

    async getCustomerById(customerId) {
        return await this.crudContext.getCustomerById(customerId);
    }

    async getAllCustomersFormatted() {
        const customers = await this.getAllCustomers();
        let result = `=== CUSTOMERS (${this.getCurrentStorageType()}) ===\n`;
        
        if (customers.length === 0) {
            result += "No customers found.\n";
        } else {
            for (const customer of customers) {
                result += `${customer.toString()}\n`;
            }
            result += `Total: ${customers.length} customers\n`;
        }
        
        return result;
    }

    async getAllCustomersFromAllFormats() {
        const formats = ["JSON", "CSV", "MongoDB"];
        let allCustomers = [];
        
        for (const format of formats) {
            this.setStorageType(format);
            const customers = await this.getAllCustomers();
            allCustomers = allCustomers.concat(
                customers.map(c => ({
                    ...c,
                    source: format
                }))
            );
        }
        
        return allCustomers;
    }
}